### PR TITLE
feat(ratelimit): qps ratelimit

### PR DESCRIPTION
**Pull request by hunshenshi**
qps ratelimit

### CODE REVIEW

File: cmd/srv-applet-mgr/apis/middleware/rate_limit.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	"context"
+
+	confrate "github.com/machinefi/w3bstream/pkg/depends/conf/rate_limit"
+	"github.com/machinefi/w3bstream/pkg/errors/status"
+)
+
+type ReqRateLimit struct{}
+
+func (r *ReqRateLimit) Output(ctx context.Context) (interface{}, error) {
+	rl, ok := confrate.RateLimitKeyFromContext(ctx)
```
Comment (huangzhiran) [inline cmd/srv-applet-mgr/apis/middleware/rate_limit.go:13]: use rl := confrate.MustRateLimitKeyFromContext(ctx)  ?

Comment (hunshenshi) [reply cmd/srv-applet-mgr/apis/middleware/rate_limit.go:13]: Then the error can't be catched.

File: cmd/srv-applet-mgr/apis/event/root.go [hunk 2]
```diff
@@ -1,11 +1,12 @@
 package event
 
 import (
+	"github.com/machinefi/w3bstream/cmd/srv-applet-mgr/apis/middleware"
 	"github.com/machinefi/w3bstream/pkg/depends/kit/httptransport"
 	"github.com/machinefi/w3bstream/pkg/depends/kit/kit"
 )
 
-var Root = kit.NewRouter(httptransport.Group("/event"))
+var Root = kit.NewRouter(httptransport.Group("/event"), &middleware.ReqRateLimit{})
```
Comment (saitofun) [inline cmd/srv-applet-mgr/apis/event/root.go:9]: EventReqRateLimit

File: pkg/depends/conf/rate_limit/rate_limit.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/machinefi/w3bstream/pkg/depends/x/misc/must"
+)
+
+type RateLimit struct {
+	Count    int            `env:""`
+	Duration types.Duration `env:""`
+
+	Limiter ratelimit.Limiter `env:""`
+}
+
```
Comment (saitofun) [inline pkg/depends/conf/rate_limit/rate_limit.go:19]: func IsZero() to check Count and Duration.

File: pkg/depends/conf/rate_limit/rate_limit.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+func (r *RateLimit) SetDefault() {}
+
+func (r *RateLimit) Init() {
+	if r.Count > 0 {
+		// TODO check Duration
+		r.Limiter = ratelimit.New(r.Count, ratelimit.Per(r.Duration.Duration()))
+	} else {
+		r.Limiter = ratelimit.NewUnlimited()
+	}
+}
```
Comment (saitofun) [inline pkg/depends/conf/rate_limit/rate_limit.go:29]: if count<=0 || duration <= 0 { unlimited }
else { new(count, duration) }

File: cmd/srv-applet-mgr/apis/middleware/rate_limit.go (hunk 1 above)
Comment (saitofun) [inline cmd/srv-applet-mgr/apis/middleware/rate_limit.go:13]: rl := mustcontext(...)
if rl.iszero() { return nil}
rl.take()...

**Review by saitofun — APPROVED** [review]

**Pull request merged**